python/cudf/cudf/tests/test_parquet.py

```diff
@@ -469,6 +470,21 @@ def test_multifile_warning(datadir):
         assert_eq(expect, got)
 
 
+# Validates the metadata return path of the parquet writer
+def test_parquet_writer_return_metadata(tmpdir, simple_gdf):
+    gdf_fname = tmpdir.join("data1.parquet")
+
+    # Write out the gdf using the GPU accelerated writer
+    df_metadata = simple_gdf.to_parquet(
+        gdf_fname.strpath, index=None, metadata_file_path="test/data1.parquet"
+    )
+    df_metadata_list = [df_metadata]
+    merged_metadata = merge_parquet_filemetadata(df_metadata_list)
+
+    # Verify that we got a valid parquet signature in the final metadata blob
+    assert merged_metadata.tobytes()[0:4] == b"PAR1"
```

Comment:
Would be nice to actually test that the metadata aggregation behaves as desired/expected:

```suggestion
    # Write out the gdf using the GPU accelerated writer
    df_metadata = simple_gdf.to_parquet(
        gdf_fname.strpath, index=None, metadata_file_path="test/data1.parquet"
    )

    # Convert metadata blob into proper metadata byte-string
    df_metadata_list_1 = [df_metadata]
    df_metadata_list_2 = [df_metadata, df_metadata]
    merged_metadata_1 = merge_parquet_filemetadata(df_metadata_list_1)
    merged_metadata_2 = merge_parquet_filemetadata(df_metadata_list_2)

    # Verify that we got a valid parquet signature in the final metadata blob
    assert merged_metadata_1.tobytes()[0:4] == b"PAR1"
    assert merged_metadata_2.tobytes()[0:4] == b"PAR1"

    # Make sure aggregation is combining metadata correctly
    fmd1 = pq.ParquetFile(BytesIO(merged_metadata_1.tobytes())).metadata
    fmd2 = pq.ParquetFile(BytesIO(merged_metadata_2.tobytes())).metadata
    assert fmd2.num_columns == fmd1.num_columns
    assert fmd2.num_rows == 2 * fmd1.num_rows
    assert fmd2.num_row_groups == 2 * fmd1.num_row_groups
```

Reply:
I can give this a shot (should work assuming that the merging doesn't verify rowgroup starting row positions). Is the pq.ParquetFile referred to here imported from pyarrow ?

Reply:
> should work assuming that the merging doesn't verify rowgroup starting row positions

Not sure I understand what you mean here?

> Is the pq.ParquetFile referred to here imported from pyarrow ?

Right, sorry: `import pyarrow.parquet as pq`

Reply:
Overall, each dask/dask_cudf worker will generate 1+ metadata objects, and then all of these objects will be sent to a single client process.  On the client process, `merge_parquet_filemetadata` will be used to convert the list of N numpy uint8_t arrays into a single "_metadata" file.  If this "_metadata" file is then opened with something like pyarrow.parquet, the number of row-groups etc must be correct (and correspond to the entire dataset).

I am now noticing that the current implementation may make it a bit complicated to write "partitioned" datasets.  It seems to be the case that the metadata returned by `to_parquet` does not include the "PAR1-magic" yet.  Is this correct? In other words, we cannot call `merge_parquet_filemetadata` on the same object multiple times, because the function is acting as bioth an aggregation and "finalizer" (please do correct me if I'm confused).   When writing partitioned datasets with pyarrow, each worker will write multiple files (for each partition) within a single task, and then locally merge it's list of metadata objects into a single object before sending along to the client.  The client will then merge a list of objects originating from all the different tasks.  I'm assuming that this all works because pyarrow does not add the "magic" until it actually writes the object to a file (it is not added during the merge).  Is it reasonable to break the "finalize" step (adding the magic bytes) into a separate function?

Reply:
Ah. That's good to know: I can change it such that the output returned by the writer also includes the header/enders so essentially identical to the output of merge.
Regarding the test and starting row positions, never mind: I thought the parquet rowgroup metadata also included a row_start parameter, but it doesn't, so shouldn't be a problem.

Reply:
Ok added the test to verify the metadata with pyarrow.parquet, and added headers to the metadata returned by to_parquet such that it's the same format than the merged output.